### PR TITLE
std: add AddressSet from rocket-chip

### PR DESCRIPTION
**Pull request by ekiwi**
### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [ ] Did you delete any extraneous printlns/debugging code?
- [ ] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [ ] Did you state the API impact?
- [ ] Did you specify the code generation impact?
- [ ] Did you request a desired merge strategy?
- [ ] Did you add text to be included in the Release Notes for this change?

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

**Comment by sequencer** [comment]
This has been implemented by `BitSet` API and `decoder.bitset` can handle this.

**Comment by sequencer** [comment]
FYI: https://github.com/chipsalliance/chisel3/blob/master/src/test/scala/chiselTests/util/BitSetSpec.scala#L95-L117

**Comment by sequencer** [comment]
I think we need a wrapper from bus library to `decoder.bitset`

**Comment by ekiwi** [comment]
How would you translate `AddressSet(0x200, 0xff)` to the `BitSet` API? Like how do you specify the offset?

**Comment by sequencer** [comment]
Basically `AddressSet` is a `BitSet` with `apply` function to convert address `offset` and `range`.
`AddressSet(0x200, 0xff)`->`BitSet(BitPat(0x200, 0xff))`

**Comment by ekiwi** [comment]
> Basically `AddressSet` is a `BitSet` with `apply` function to convert address `offset` and `range`.

I see. So you can basically think of `value` as being and `offset`. Do you know if there is a good way to expose the mask to the user?

I am trying to see how `BitSet` could be used with the [AXI4 SRAM](https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/amba/axi4/SRAM.scala) from rocket chip. Things that are missing so far are:
- [`.mask`](https://github.com/chipsalliance/rocket-chip/blob/b36060ebe7bd821ea4b4d1dd2aa9e6d38e7d01b4/src/main/scala/diplomacy/SRAM.scala#L31)
- [`.contains(UInt): Bool`](https://github.com/chipsalliance/rocket-chip/blob/b36060ebe7bd821ea4b4d1dd2aa9e6d38e7d01b4/src/main/scala/amba/axi4/SRAM.scala#L67)

**Comment by sequencer** [comment]
RC Address Decoder use a heuristic algorithm to solve the partition problem, however the PPA result is bad, from evaluation at my side, decoder API provides about 30% PPA improvement in some case, so I directly drop it.
I think what we need here is a factory to convert address and address range to `BitSet`.

**Comment by ekiwi** [comment]
> Yes, that's the key issue, we need a address factory, to convert any `AddressSet` to `BitSet`.

But then we would still need `AddressSet`. I thought the whole point was to reduce all uses of `AddressSet` with `BitSet` in order to avoid duplicate functionality.

**Comment by sequencer** [comment]
> > Yes, that's the key issue, we need a address factory, to convert any `AddressSet` to `BitSet`.
> 
> But then we would still need `AddressSet`. I thought the whole point was to reduce all uses of `AddressSet` with `BitSet` in order to avoid duplicate functionality.

Yes, I do think so, we need a new API and implementations for `AddressSet`.
`mask` is trivial.
`contains` should be implemented via decoder. I must forgot to implement that.
I'll add it to `BitSet`, and `AddressSet` can directly wrap that.

**Comment by sequencer** [comment]
Just thought about the `contains`, which is trivial as well:
```
def contains(data: UInt): decoder.bitset(data, Seq(this))
```
`this` is the `BitSet` need to test on the contains.

File: stdlib/src/chisel3/std/AddressSet.scala [hunk 1]
```diff
@@ -0,0 +1,134 @@
+package chisel3.std
+
+import chisel3._
+import chisel3.util.log2Floor
+
+/** AddressSets specify the address space managed by the manager
```
Comment (mwachs5) [inline stdlib/src/chisel3/std/AddressSet.scala:6]: If this is getting upstreamed it needs solid scaladoc on each public function

**Comment by sequencer** [comment]
Note: there is a big difference between `BitSet` and RC `AddressSet`:
RC `AddressSet` can be unknown width, as you can see from `val Everything = AddressSet(0, -1)` API, while chisel BitSet doesn't support.

There are a lot of useful APIs for interconnection library, some are covered by `BitSet`, while some are not:
- `def contains(x: BigInt): Boolean` can be replaced by `BitSet.cover`
- `def contains(x: UInt): Bool` is not implemented, but I think we can implement by `decoder.BitSet` with a more general API, which can test a hard signal can be multiple `AddressSet` covered or not, this will significantly improve the PPA.
- `def legalize(x: UInt): UInt` can be dropped.
- `def overlaps(x: AddressSet)` is `def overlap(that: BitSet)`
- `def contains(x: AddressSet)` is `def cover(that: BitSet)`
- `def alignment: BigInt` is a useful helper, but it should not exist in `BitSet`
- `def contiguous: Boolean` is a useful helper, but it should not exist in `BitSet`
- `def finite: Boolean` is dropped, `BitSet` must be known width.
- `def max: BigInt` is a useful helper, but it should not exist in `BitSet`
- `def widen(imask: BigInt)` IIRC, is only useful in some test, I think we should keep the API simple.
- `def intersect(x: AddressSet)` is `def intersect(that: BitSet): BitSet`
- `def subtract(x: AddressSet) ` is `def subtract(that: BitSet): BitSet`
- `def compare(x: AddressSet): Int` is useless, IIRC, is is for the ordering, which used in address decoder.
- `def toString`, we certainly need this!

most API in `object AddressSet` is designed for decoder, I think they can safely dropped.

**Pull request closed without merging**